lib/block.js

```diff
@@ -0,0 +1,147 @@
+    // NOTE: should there be a separation between context as something global
+    // vs. context as quilt/tailor-specific term?
+    let context = originalContext;
+
+    const resultStream = new StringifierStream((tag) => {
+
+        const {
+            maxAssetLinks,
+            asyncStream,
+            handleTag,
+            requestFragment,
+
+            // Should the two below be one???
+            pipeDefinition,
+            pipeAttributes,
+            pipeInstanceName,
+
+
+            dynamicContextAttribute,
+            fetchDynamicContext
+        } = context;
+
+        const { placeholder, name } = tag;
+
+        const fetchFragment = (combinedContext) => {
+            const fragment = new Fragment({
+                tag,
+                context: combinedContext,
+                index: combinedContext.index,
+                requestFragment,
+                pipeInstanceName,
+                pipeAttributes,
+                maxAssetLinks
+            });
+
+            FRAGMENT_EVENTS.forEach((eventName) => {
+                fragment.on(eventName, (...args) => {
+                    const prefixedName = 'fragment:' + eventName;
+                    resultStream.emit(prefixedName, request, fragment.attributes, ...args);
+                });
+            });
+
+            const isAsync = fragment.attributes.async;
+            const isPrimary = fragment.attributes.primary;
+
+            if (isAsync) {
+                asyncStream.write(fragment.stream);
+            }
+
+            if (isPrimary) {
+                resultStream.emit('primary:found', fragment);
+            }
+
+            return fragment.fetch(request, false);
+        };
+
+        if (placeholder === 'pipe') {
+            return pipeDefinition(pipeInstanceName);
+        }
+
+        if (asyncStream && placeholder === 'async') {
+            // end of body tag
+            return asyncStream;
+        }
+
+        if (name === 'fragment') {
+            // Freeze current context & update the indexes
+            const fragmentContext = context;
+            context = Object.assign({}, context, {
+                index: context.index + maxAssetLinks
+            });
+            resultStream.emit('fragment:found');
+
+            const shouldFetchDynamicContext = dynamicContextAttribute
+                && tag.attributes
+                && (typeof tag.attributes[dynamicContextAttribute] !== 'undefined');
+
+            if (shouldFetchDynamicContext) {
+                const tempStream = new PassThrough({
+                    objectMode: true
+                });
+
+                const pipeFragmentResult = result => {
+                    if (result instanceof stream) {
+                        result.pipe(tempStream);
+                    } else {
+                        tempStream.end(result);
+                    }
+                };
+
+                fetchDynamicContext(tag)
```

Comment:
IMO this creates unnecessary complexity in tailor itself and can be handled outside in requestFragment logic. 

Tailor itself is extendable, we may not need this feature here.

Reply:
Good thoughts. It doesn't really belong to this PR.